java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java

```diff
@@ -0,0 +1,152 @@
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.Iterator;
+
+import org.apache.arrow.adapter.jdbc.consumer.CompositeJdbcConsumer;
+import org.apache.arrow.adapter.jdbc.consumer.JdbcConsumer;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+/**
+ * VectorSchemaRoot iterator for partially converting JDBC data.
+ */
+public class ArrowVectorIterator implements Iterator<VectorSchemaRoot> {
+
+  private final ResultSet resultSet;
+  private final JdbcToArrowConfig config;
+
+  private final Schema schema;
+  private final ResultSetMetaData rsmd;
+
+  private final JdbcConsumer[] consumers;
+  private final CompositeJdbcConsumer compositeConsumer;
+  private boolean consumerCreated = false;
+
+  private VectorSchemaRoot nextBatch;
+
+  private final int partialLimit;
+
+  /**
+   * Construct an instance.
+   */
+  private ArrowVectorIterator(ResultSet resultSet, JdbcToArrowConfig config) throws SQLException {
+    this.resultSet = resultSet;
+    this.config = config;
+    this.schema = JdbcToArrowUtils.jdbcToArrowSchema(resultSet.getMetaData(), config);
+    this.partialLimit = config.getPartialLimit();
+
+    rsmd = resultSet.getMetaData();
+    consumers = new JdbcConsumer[rsmd.getColumnCount()];
+    this.compositeConsumer = new CompositeJdbcConsumer(consumers);
+  }
+
+  /**
+   * Create a ArrowVectorIterator to partially convert data.
+   */
+  public static ArrowVectorIterator create(
+      ResultSet resultSet,
+      JdbcToArrowConfig config)
+      throws SQLException {
+
+    ArrowVectorIterator iterator = new ArrowVectorIterator(resultSet, config);
+    try {
+      iterator.load();
+      return iterator;
+    } catch (Exception e) {
+      iterator.close();
+      throw new RuntimeException("error occurs while creating iterator", e);
+    }
+  }
+
+  // Loads the next schema root or null if no more rows are available.
+  private void load() throws IOException, SQLException {
+    VectorSchemaRoot root = VectorSchemaRoot.create(schema, config.getAllocator());
+    JdbcToArrowUtils.allocateVectors(root, JdbcToArrowUtils.DEFAULT_BUFFER_SIZE);
+
+    if (!consumerCreated) {
+      consumerCreated = true;
+      for (int i = 1; i <= consumers.length; i++) {
+        consumers[i - 1] = JdbcToArrowUtils.getConsumer(resultSet, i, resultSet.getMetaData().getColumnType(i),
+            root.getVector(rsmd.getColumnName(i)), config);
+      }
+    } else {
+      for (int i = 1; i <= consumers.length; i++) {
+        consumers[i - 1].resetValueVector(root.getVector(rsmd.getColumnName(i)));
+      }
+    }
+
+    // consume data
+    try {
+      int readRowCount = 0;
+      while (resultSet.next()) {
+        compositeConsumer.consume(resultSet);
+        readRowCount++;
+        if (partialLimit != JdbcToArrowConfig.DEFAULT_PARTIAL_LIMIT && readRowCount >= partialLimit) {
+          break;
+        }
```

Comment:
can the condition be moved into the while?

Reply:
Nope, like while(rs.next() && condition) will has some problem
if rs.next() = true and condition = false, the result cursor will move ahead and we will lose data. (If ResultSet has API like hasNext without cursor moving, we can do like this)

Reply:
what about `while (condition && rs.next())`